components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -0,0 +1,406 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::metrics::*;
+use super::IOStats;
+use crate::IOType;
+
+use collections::HashMap;
+use std::collections::VecDeque;
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+use bcc::{table::Table, Kprobe, BPF};
+use crossbeam_utils::CachePadded;
+
+/// Biosnoop leverages BCC to make use of eBPF to get disk IO of TiKV requests.
+/// The BCC code is in `biosnoop.c` which is compiled and attached kernel on
+/// TiKV bootstrap. The code hooks on the start and completion of blk_account_io
+/// in kernel, so it's easily to get the latency and bytes of IO requests issued
+/// by current PID.
+///
+/// The main usage of iosnoop is to get accurate disk IO of different tasks
+/// separately, like compaction, coprocessor and raftstore, instead of a global
+/// disk throughput. So IO-types should be tagged for different threads by
+/// `set_io_type()`. And BCC code is available to get the IO-type for one thread
+/// by address, then all the IO requests for that thread will be recorded in
+/// corresponding type's map in BCC.
+///
+/// With that information, every time calling `IOContext` it get the stored stats
+/// from corresponding type's map in BCC. Thus it enables TiKV to get the latency and
+/// bytes of read/write request per IO-type.
+
+const MAX_THREAD_IDX: usize = 192;
+
+// Hold the BPF to keep it not dropped.
+// The two tables are `stats_by_type` and `type_by_pid` respectively.
+static mut BPF_TABLE: Option<(BPF, Table, Table)> = None;
+
+// This array records the io type for every thread. The address of this array
+// will be passed into BPF, so BPF code can get io type for specific thread
+// without an extra syscall.
+// It should be a thread local variable, but the address of thread local is not
+// reliable. So define a global array and let each thread writes on a specific
+// element. Thus there is no contention for the element and use padding to avoid
+// false sharing.
+// Leave the last element as reserved, when there is no available index, all
+// other threads will be allocated to that index with IOType::Other always.
+static mut IO_TYPE_ARRAY: [CachePadded<IOType>; MAX_THREAD_IDX + 1] =
```

Comment:
<img width="825" alt="WeChatWorkScreenshot_5c4b912b-ca66-411f-aee4-0e3c43aaa71a" src="https://user-images.githubusercontent.com/2606959/102838483-2726e700-43b3-11eb-97d2-5bfe400da66c.png">

but it is also being access in bpf, supposed in kernel thread?

Reply:
In BCC code, the io type is read when `blk_account_io_start` is called which is fired in the process context. That is to say, it's called in kernel space of user thread rather than kernel thread. So there is no contention between user and kernel.

Reply:
got it. so much to learn here. Can you add comment to explain why it doesn't have to be atomic?